### ISSUE TITLE
Optimization Verification test

### ISSUE THREAD

**Issue by tomkocmi** [ufal/neuralmonkey#745]
Is it possible to compute optimization verification test?
When you are translating a sentence from devset and the translation is different than reference sentence, I would like to find out if the problem is due to a beam search, who did not manage to find the correct translation. Or if it is due to the model, which evaluates correct translation with a worse score than the decoded translation.

**Comment by jindrahelcl** [comment]
I don't know why but I've rewrited this answer at least three times always with a different answer. At the end, I settled with this:

I think this could be done with something as the perplexity runner. You provide the sequences you want to score as the targets to the decoder and then measure the runtime (or train) loss on these targets.
For this, you would need to (copy and) adapt the perplexity runner (which now works only with classifiers) to work with autoregressive decoders as well (i.e. objects that generate sequences).

To conclude, no, I don't think huge changes are necessary. I think most of the functionality is already there.

**Comment by jindrahelcl** [comment]
EDIT: Of course the perplexity runner does already work with autoregressive decoders! I was just confused from my previous thoughts

**Comment by jindrahelcl** [comment]
So, solution: First, translate your data. Second, in the config, lose the runners generating the sentences and replace them with the perplexity runner which adds a new output data series which gets written into a file. Third, run the config twice, a) with reference sentences as `s_target`, b) with your translations as `s_target`. Fourth, compare the numbers in the two outputting files (they should have a number on each line and should have an equal number of lines as your input files).

**Comment by jindrahelcl** [comment]
... if you want to, change `train_xents` to `runtime_xents` in the perplexity runner and you will get another set of observations to inspect.

**Comment by varisd** [comment]
I think, you want to use runtime_xents if you want to know the probability (in this case perplexity) of the decoded sentence (train_xents uses training mode - each word generated in a time step is conditioned on the gold output from the previous step - the resulting probability/perplexity is not the one of the decoded sequence).

**Comment by jindrahelcl** [comment]
There is no "gold" input in this setup - you can replace the reference data with anything it will get fed into the decoder in training mode. There can be some interesting differences and maybe you can measure the exposure bias using these. But of course, runtime logits alone are perhaps more relevant than training logits.
